File: MyApi/Services/PdfService.cs
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;
using MyApi.Models;

namespace MyApi.Services;

public class PdfService
{
    static PdfService()
    {
        QuestPDF.Settings.License = LicenseType.Community;
    }

    public byte[] GenerateOrderPdf(Request request)
    {
        var document = Document.Create(container =>
        {
            container.Page(page =>
            {
                page.Size(PageSizes.A4);
                page.Margin(30);
                page.DefaultTextStyle(x => x.FontSize(11));

                page.Header().Element(c => ComposeHeader(c, request));
                page.Content().Element(c => ComposeContent(c, request));
                page.Footer().Element(ComposeFooter);
            });
        });

        return document.GeneratePdf();
    }

    private void ComposeHeader(IContainer container, Request request)
    {
        container.Row(row =>
        {
            row.RelativeItem().Column(column =>
            {
                column.Item().Text("ORDRE DE PRÉLÈVEMENT").Bold().FontSize(18).FontColor(Colors.Blue.Darken2);
                column.Item().Text($"N°: ORD-{request.Id:D6}").Bold().FontSize(14);
            });

            row.ConstantItem(100).Background(Colors.Blue.Lighten5).Padding(10).Column(column =>
            {
                column.Item().Text("Économat").Bold().FontSize(12);
                column.Item().Text("Date: " + DateTime.Now.ToString("dd/MM/yyyy")).FontSize(10);
            });
        });

        container.PaddingVertical(10);
    }

    private void ComposeContent(IContainer container, Request request)
    {
        container.Column(column =>
        {
            // Request Info Section
            column.Item().Background(Colors.Grey.Lighten4).Padding(10).Row(row =>
            {
                row.RelativeItem().Column(col =>
                {
                    col.Item().Text("Service demandeur:").Bold();
                    col.Item().Text(request.ServiceName);
                });
                
                row.RelativeItem().Column(col =>
                {
                    col.Item().Text("Date de la demande:").Bold();
                    col.Item().Text(request.RequestDate.ToString("dd/MM/yyyy HH:mm"));
                });
                
                row.RelativeItem().Column(col =>
                {
                    col.Item().Text("Statut:").Bold();
                    col.Item().Text("FINALISÉ ✅").FontColor(Colors.Green.Darken2);
                });
            });

            column.Item().PaddingVertical(10);

            // Items Table
            column.Item().Text("Articles à prélever:").Bold().FontSize(13);
            column.Item().PaddingTop(5).Table(table =>
            {
                table.ColumnsDefinition(columns =>
                {
                    columns.ConstantColumn(40);
                    columns.RelativeColumn(3);
                    columns.RelativeColumn();
                    columns.RelativeColumn();
                    columns.RelativeColumn(2);
                });

                table.Header(header =>
                {
                    header.Cell().Background(Colors.Blue.Darken2).Padding(5).Text("N°").FontColor(Colors.White).Bold();
                    header.Cell().Background(Colors.Blue.Darken2).Padding(5).Text("Article demandé").FontColor(Colors.White).Bold();
                    header.Cell().Background(Colors.Blue.Darken2).Padding(5).Text("Code Produit").FontColor(Colors.White).Bold();
                    header.Cell().Background(Colors.Blue.Darken2).Padding(5).Text("Désignation").FontColor(Colors.White).Bold();
                    header.Cell().Background(Colors.Blue.Darken2).Padding(5).Text("Quantité").FontColor(Colors.White).Bold();
                });

                var index = 1;
                foreach (var item in request.Items)
                {
                    var bgColor = index % 2 == 0 ? Colors.White : Colors.Grey.Lighten4;
                    
                    table.Cell().Background(bgColor).Padding(5).Text(index.ToString());
                    table.Cell().Background(bgColor).Padding(5).Text(item.ProductName);
                    table.Cell().Background(bgColor).Padding(5).Text(item.Product?.Code_Produit ?? "-");
                    table.Cell().Background(bgColor).Padding(5).Text(item.Product?.Designation ?? "-");
                    table.Cell().Background(bgColor).Padding(5).Text($"{item.Quantity} {item.Unit}");
                    
                    index++;
                }
            });

            column.Item().PaddingVertical(20);

            // Signature Section
            column.Item().Row(row =>
            {
                row.RelativeItem().Column(col =>
                {
                    col.Item().Text("Signature du responsable économat:").FontSize(10);
                    col.Item().PaddingTop(30).BorderTop(1).Text("");
                });

                row.RelativeItem().Column(col =>
                {
                    col.Item().Text("Signature du demandeur:").FontSize(10);
                    col.Item().PaddingTop(30).BorderTop(1).Text("");
                });
            });
        });
    }

    private void ComposeFooter(IContainer container)
    {
        container.Row(row =>
        {
            row.RelativeItem().Text("Généré le: " + DateTime.Now.ToString("dd/MM/yyyy HH:mm")).FontSize(9).FontColor(Colors.Grey.Medium);
            row.RelativeItem().AlignRight().Text("Page x of y").FontSize(9).FontColor(Colors.Grey.Medium);
        });
    }
}

File: client/src/services/pdfService.js
import jsPDF from 'jspdf';

export const downloadPdf = (request) => {
  try {
    const doc = new jsPDF();
    
    // Header
    doc.setFontSize(18);
    doc.setTextColor(0, 51, 102);
    doc.text('ORDRE DE PRÉLÈVEMENT', 105, 20, { align: 'center' });
    
    doc.setFontSize(12);
    doc.setTextColor(0, 0, 0);
    doc.text(`N°: ORD-${String(request.id).padStart(6, '0')}`, 20, 35);
    
    // Date box
    doc.setFillColor(220, 230, 241);
    doc.rect(150, 25, 40, 15, 'F');
    doc.setFontSize(10);
    doc.text('Économat', 155, 32);
    doc.text(`Date: ${new Date().toLocaleDateString('fr-FR')}`, 155, 38);
    
    // Request info
    doc.setFillColor(240, 240, 240);
    doc.rect(20, 45, 170, 25, 'F');
    
    const serviceName = request.serviceName || '-';
    const requestDate = request.requestDate ? new Date(request.requestDate).toLocaleString('fr-FR') : '-';
    
    doc.setFontSize(10);
    doc.text('Service demandeur:', 25, 55);
    doc.setFontSize(11);
    doc.setFont('helvetica', 'bold');
    doc.text(serviceName, 25, 62);
    
    doc.setFont('helvetica', 'normal');
    doc.setFontSize(10);
    doc.text('Date de la demande:', 80, 55);
    doc.setFontSize(11);
    doc.setFont('helvetica', 'bold');
    doc.text(requestDate, 80, 62);
    
    doc.setFont('helvetica', 'normal');
    doc.setFontSize(10);
    doc.text('Statut:', 135, 55);
    doc.setFontSize(11);
    doc.setTextColor(0, 100, 0);
    doc.setFont('helvetica', 'bold');
    doc.text('FINALISÉ', 135, 62);
    doc.setTextColor(0, 0, 0);
    
    // Table with all columns: Article, Code Produit, Désignation, Quantité
    doc.setFontSize(12);
    doc.setFont('helvetica', 'bold');
    doc.text('Articles:', 20, 82);
    
    let yPos = 90;
    doc.setFillColor(0, 51, 102);
    doc.rect(20, yPos, 170, 8, 'F');
    doc.setTextColor(255, 255, 255);
    doc.setFontSize(9);
    doc.text('Article', 22, yPos + 5);
    doc.text('Code', 75, yPos + 5);
    doc.text('Désignation', 105, yPos + 5);
    doc.text('Qté', 175, yPos + 5);
    
    yPos += 10;
    doc.setTextColor(0, 0, 0);
    doc.setFont('helvetica', 'normal');
    
    const items = request.items || [];
    items.forEach((item, index) => {
      const bgColor = index % 2 === 0 ? 245 : 255;
      doc.setFillColor(bgColor, bgColor, bgColor);
      doc.rect(20, yPos - 2, 170, 8, 'F');
      
      const article = item.productName || '-';
      const code = item.productCode || item.product?.code_Produit || '-';
      const designation = item.productDesignation || item.product?.designation || '-';
      const quantite = `${item.quantity || 0} ${item.unit || ''}`;
      
      doc.text(String(article).substring(0, 20), 22, yPos + 3);
      doc.text(String(code).substring(0, 12), 75, yPos + 3);
      doc.text(String(designation).substring(0, 30), 105, yPos + 3);
      doc.text(quantite, 175, yPos + 3);
      
      yPos += 8;
    });
    
    // Signatures in a table
    yPos += 15;
    
    doc.setFontSize(12);
    doc.setFont('helvetica', 'bold');
    doc.text('Signatures:', 20, yPos);
    
    yPos += 8;
    doc.setFillColor(0, 51, 102);
    doc.rect(20, yPos, 170, 8, 'F');
    doc.setTextColor(255, 255, 255);
    doc.setFontSize(10);
    doc.text('Demandeur', 35, yPos + 5);
    doc.text('Control', 75, yPos + 5);
    doc.text('Directeur', 115, yPos + 5);
    doc.text('Economat', 155, yPos + 5);
    
    yPos += 10;
    doc.setTextColor(0, 0, 0);
    doc.setFont('helvetica', 'normal');
    doc.setFontSize(11);
    doc.setFillColor(245, 245, 245);
    doc.rect(20, yPos - 2, 170, 15, 'F');
    
    doc.text('Signé', 35, yPos + 3);
    doc.text('Signé', 75, yPos + 3);
    doc.text('Signé', 115, yPos + 3);
    doc.text('Signé', 155, yPos + 3);
    
    // Footer
    doc.setFontSize(9);
    doc.setTextColor(128, 128, 128);
    doc.text(`Généré le: ${new Date().toLocaleString('fr-FR')}`, 20, 280);
    
    // Save
    doc.save(`OrdrePrelevement_${request.id}.pdf`);
  } catch (err) {
    console.error('PDF generation error:', err);
    throw err;
  }
};